File: Exp1_CommNoises.py
import numpy as np
import matplotlib.pyplot as plt

# Simulate a clean signal (a simple sine wave)
fs = 1000  # Sampling frequency
t = np.linspace(0, 1, fs)  # Time vector
freq = 5  # Frequency of the sine wave
clean_signal = np.sin(2 * np.pi * freq * t)

# 1. Gaussian Noise
mean = 0
variance = 0.5  # Noise variance
gaussian_noise = np.random.normal(mean, np.sqrt(variance), fs)
noisy_signal_gaussian = clean_signal + gaussian_noise

# 2. Impulse Noise
impulse_noise = np.zeros(fs)
impulse_positions = np.random.randint(0, fs, size=50)  # Random positions for impulses
impulse_noise[impulse_positions] = np.random.choice([-5, 5], size=50)  # Impulse values
noisy_signal_impulse = clean_signal + impulse_noise

# 3. White Noise
white_noise = np.random.uniform(-1, 1, fs)  # Uniform distribution for white noise
noisy_signal_white = clean_signal + white_noise

# Plot the clean signal and the noisy signals
plt.figure(figsize=(10, 12))

# Clean Signal Plot
plt.subplot(4, 1, 1)
plt.plot(t, clean_signal)
plt.title("Clean Signal (Sine Wave)")

# Noisy Signal with Gaussian Noise
plt.subplot(4, 1, 2)
plt.plot(t, noisy_signal_gaussian)
plt.title("Noisy Signal with Gaussian Noise")

# Noisy Signal with Impulse Noise
plt.subplot(4, 1, 3)
plt.plot(t, noisy_signal_impulse)
plt.title("Noisy Signal with Impulse Noise")

# Noisy Signal with White Noise
plt.subplot(4, 1, 4)
plt.plot(t, noisy_signal_white)
plt.title("Noisy Signal with White Noise")

plt.tight_layout()
plt.show()

# Calculate Signal-to-Noise Ratio (SNR) for each type of noise
snr_gaussian = 10 * np.log10(np.var(clean_signal) / np.var(gaussian_noise))
snr_impulse = 10 * np.log10(np.var(clean_signal) / np.var(impulse_noise))
snr_white = 10 * np.log10(np.var(clean_signal) / np.var(white_noise))

print(f"Signal-to-Noise Ratio (SNR) with Gaussian Noise: {snr_gaussian:.2f} dB")
print(f"Signal-to-Noise Ratio (SNR) with Impulse Noise: {snr_impulse:.2f} dB")
print(f"Signal-to-Noise Ratio (SNR) with White Noise: {snr_white:.2f} dB")

File: Exp1_CommNoises.m
clc;
clear;
close all;

% Simulate a clean signal (sine wave)
fs = 1000;                     % Sampling frequency
t = linspace(0, 1, fs);        % Time vector
freq = 5;                      % Frequency of sine wave
clean_signal = sin(2*pi*freq*t);

% 1. Gaussian Noise
mean_val = 0;
variance = 0.5;
gaussian_noise = mean_val + sqrt(variance) * randn(1, fs);
noisy_signal_gaussian = clean_signal + gaussian_noise;

% 2. Impulse Noise
impulse_noise = zeros(1, fs);
impulse_positions = randi([1 fs], 1, 50);
impulse_values = 5 * (2*randi([0 1], 1, 50) - 1); % Generates -5 or 5
impulse_noise(impulse_positions) = impulse_values;
noisy_signal_impulse = clean_signal + impulse_noise;

% 3. White Noise
white_noise = -1 + 2*rand(1, fs);  % Uniform distribution between -1 and 1
noisy_signal_white = clean_signal + white_noise;

% Plotting
figure;

subplot(4,1,1);
plot(t, clean_signal, 'LineWidth', 1.2);
title('Clean Signal (Sine Wave)');
xlabel('Time');
ylabel('Amplitude');
grid on;

subplot(4,1,2);
plot(t, noisy_signal_gaussian, 'LineWidth', 1.2);
title('Noisy Signal with Gaussian Noise');
xlabel('Time');
ylabel('Amplitude');
grid on;

subplot(4,1,3);
plot(t, noisy_signal_impulse, 'LineWidth', 1.2);
title('Noisy Signal with Impulse Noise');
xlabel('Time');
ylabel('Amplitude');
grid on;

subplot(4,1,4);
plot(t, noisy_signal_white, 'LineWidth', 1.2);
title('Noisy Signal with White Noise');
xlabel('Time');
ylabel('Amplitude');
grid on;

% SNR Calculation
snr_gaussian = 10 * log10(var(clean_signal) / var(gaussian_noise));
snr_impulse  = 10 * log10(var(clean_signal) / var(impulse_noise));
snr_white    = 10 * log10(var(clean_signal) / var(white_noise));

fprintf('Signal-to-Noise Ratio (SNR) with Gaussian Noise: %.2f dB\n', snr_gaussian);
fprintf('Signal-to-Noise Ratio (SNR) with Impulse Noise: %.2f dB\n', snr_impulse);
fprintf('Signal-to-Noise Ratio (SNR) with White Noise: %.2f dB\n', snr_white);